Optimistic concurrency check on update

cdonnellytx opened the pull request:
Adding support for doing optimistic concurrency checks on update using the [System.ComponentModel.DataAnnotations](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations) attributes:

* [ConcurrencyCheck](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.concurrencycheckattribute) - used for manually modified fields
* [Timestamp](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.timestampattribute) - used for fields modified by the database on insert/update.  This includes intrinsic field types like [rowversion](https://docs.microsoft.com/en-us/sql/t-sql/data-types/rowversion-transact-sql) as well as fields that simulate this by using database triggers.

## How it works

The attributes are checked on `Update<T>` object calls (both singular and list, sync and async) by adding the additional fields to the `WHERE` clause of the `UPDATE` statement.

If the values match, the update goes through; if the values don't match, an `OptimisticConcurrencyException` is thrown.

We had to include our own version of the exception because while .NET Framework does have an [OptimisticConcurrencyException](https://docs.microsoft.com/en-us/dotnet/api/system.data.optimisticconcurrencyexception?view=netframework-4.8&viewFallbackFrom=netcore-3.1), it is not in .NET Core outside of EF Core (where it has a different name).

Tests cover concurrency checks on both updates and upserts (which follow from updates).

## Other changes

Changes I made that are only tangentially related to this change:

* Tweak to .editorconfig because Vim's support for .editorconfig doesn't support `tabs`, only `tab`.
* We can now run PostgreSQL tests on a fresh database (the static initializer catches the appropriate exception, runs a `create database test`, then reruns the initializer script).
* Renamed the 7 `cmd` variables to match the 18 `command` variables in the SqlAdapter files to try to see how many of the various `UPDATE` implementations actually do the same thing.  (Not enough to merge.)
* Reformatting headers in the attributes section in the README to be a bit clearer (in addition to docs on the new attributes).

cdonnellytx committed:
Simplifying attribute check expressions.

The attribute expressions were repetitive (multiple calls to GetCustomAttributes(false)), and some calls were for the same attributes (ReadOnlyAttribute, DatabaseGeneratedAttribute).

This change reduces the number of GetCustomAttributes() calls to two, and the calls on the array of attributes now use more concise LINQ methods that avoid lambdas in most cases (not avoidable for the dynamic attributes, however).

This is in preparation for adding optimistic concurrency, which checks more attributes.

cdonnellytx committed:
TableInfo now tracks DataAnnotations concurrency attributes.

cdonnellytx committed:
[DOC] Tweaking header format.

* Adding an overall "attributes" header
* Turning the attribute library names into subheaders
* Turning the attribute names into H4 headers, and links where applicable.
* Adding a subheader for Example

cdonnellytx committed:
[DOC] Adding docs for ConcurrencyCheck and Timestamp attributes.

cdonnellytx committed:
Adding UPDATE concurrency check code

cdonnellytx committed:
EditorConfig officially only supports "tab" not "tabs".

A few editors out there don't support "tabs", though this may be changing:
https://github.com/editorconfig/editorconfig/issues/363

cdonnellytx committed:
Convert SQL to UTF-8 so it can be diffed.

cdonnellytx committed:
Adding a RowVersion column to test TimestampAttribute

cdonnellytx committed:
Adding a ComparisonColumns member alongside ConcurrencyCheckColumns.

* It means we effectively only ever union key/check columns once, at lazy-eval time.  After that the enumerable is constructed.
* We still need separate ConcurrencyCheckColumns for detecting failures later (so we can throw an exception).
* Adding TimestampAttribute test to SQL Server only.  While it is possible to do in some other databases, such as Oracle, the mechanism differs (typically done via a trigger), and cannot be tested in the same way.

cdonnellytx committed:
Adding upsert concurrency check tests.

cdonnellytx committed:
Adding concurrency tests for UpdateList.

Note that the way it currently works, if item K/N fails, items 1..K-1 are modified, and K..N are unmodified.

cdonnellytx committed:
Adding exception for optimistic concurrency failure plus tests.

Throwing an exception is the ideal outcome here; concurrency checks should rarely fail, and if they do, it should be up to the caller to resolve them.

This should also resolve the problem with bulk operations where some items were changed and some were not: *all* items in the statement should be unchanged.

cdonnellytx committed:
Wiring up concurrency check for sync updates.

cdonnellytx committed:
UpdateAsync now throws OptimisticConcurrencyException

cdonnellytx committed:
Splitting SqlServerTestSuite into 3 files: base, sync, async.

By doing this I found a sync test without an async partner; adding it.

cdonnellytx committed:
Fixing Timestamp modified tests to catch exception.

cdonnellytx committed:
Fixing discrepancy between "cmd" and "command".

"command" is more prevalent.

cdonnellytx committed:
Propagating commandTimeout into the concurrency check.

cdonnellytx committed:
Making SqlAdapter.Update|UpdateAsync<T>() do a concurrency check.

Doing this by moving the actual UPDATE statements to new methods ("UpdateInternal", "UpdateInternalAsync") and making the base class do the concurrency checks.

This simplifies things greatly, but still allows for any subclasses to override the entirety of the Update/UpdateAsync methods.

cdonnellytx committed:
We can now run PostgreSQL tests on a fresh database.

I finally figured out how to fall back and create the test database.

cdonnellytx committed:
Adding async update tests.

cdonnellytx committed:
Removing separate ConcurrencyCheck methods.

They're now just part of the Update methods.

HudsonAkridge commented on Dapper.Tests.Database/Tests/TestSuiteUpdate.cs:
```diff
@@ -187,5 +188,32 @@ public void UpdatePartial()
+
+        [Fact]
+        [Trait("Category", "Update")]
+        public void UpdateConcurrencyCheck()
```
We're testing for the failed use case with the `[ConcurrencyCheck]` attribute, but not the success where it doesn't fail on a field marked with `[ConcurrencyCheck]`, correct?

For all the tests testing update, we should make sure that it properly allows for an update when the attribute is used in the WHERE query generator and the row was *not* updated in another connection context.

I wrote a quick test copy of this that did just that, and it succeeds, but would be good to add them for all updates for the explicit behavior we want.

Additionally, there's no limits on specifying multiple `[ConcurrencyCheck]` columns in your model from what I can see, so we should consider adding a model with two properties using that attribute to ensure that we don't need to enforce only specifying it once on a model. (As I understand the way you built/wrote this, this shouldn't be an issue, it should just work, but we should consider adding that test case since we're not enforcing only having the attribute specified once on the model)

cdonnellytx replied:
The updates have a single test for concurrency success and failure, but I'll go ahead and split the tests to make it clearer.

The main reason I avoided doing multiple columns was to avoid changing the tables as much as possible, but I'll go ahead and do that so the tests cover that.

cdonnellytx replied:
OK, concurrency checks tests have been rewritten to have separate modified/not-modified tests for Update, UpdateList, Upsert sync/async variants.

While doing this I uncovered a bug wherein we weren't properly comparing nullable ConcurrencyCheck fields, as the SQL generated did not handle null semantics.  So I added metadata to ColumnInfo to tell if a field was nullable.

The following determines whether a field is nullable or not:

* Keys (those marked with [KeyAttribute]) are not null.
* Non-keys marked with [RequiredAttribute] are not null.
* Otherwise, nullness is determined by data type:
    * Value types are not nullable unless they are of type `Nullable<T>`.
    * For reference types, the new C# 8.0 [nullable references](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references) are honored in a similar fashion when nullable checks are enabled (more formally, when they are within a nullable annotation context).
    * When nullable checks are not enabled (C# 7.3 or earlier, or `#nullable disable`), reference types are considered nullable.

[KeyAttribute]: https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.keyattribute
[RequiredAttribute]: https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.requiredattribute

HudsonAkridge reviewed (changes requested):
This looks good, just needs those tests added for the positive use case.

Implementation seems sound, good job. This was a lot cleaner than the road I began going down a few years ago :)

cdonnellytx committed:
Changing concurrency check model to mark multiple fields.

Unfortunately this revealed a bug: we aren't handling nulls in comparisons.  Need to know if a field is nullable or not...

cdonnellytx committed:
Adding support for checking if a property is nullable.

Tests cover nullable structs as well as the new C# 8 nullable reference types.

cdonnellytx committed:
Modifying SQL generation to handle nullable columns.

* ColumnInfo does not allow key columns to be null both for backward compatibility and for performance reasons.
* WHERE clause comparison now performs both equality and null checks if the column is nullable.

cdonnellytx committed:
Splitting UpdateConcurrencyCheck into two separate tests.

This is to better illustrate the unmodified versus modified tests.

Additionally, it now covers that the concurrency check works just fine with nullable fields.

cdonnellytx committed:
Ensuring all other concurrency checks have separate (not)modified tests.

cdonnellytx committed:
Adding partial support for RequiredAttribute.

This will allow pre-C# 8.0 code to mark columns as not null.

Note that we do **not** honor the "no empty/blank value" semantics.

cdonnellytx committed:
Adding tests to cover ColumnInfo.IsNullable.

cdonnellytx committed:
Adding RequiredAttribute to docs.

Also explaining nullable checks also work.

cdonnellytx committed:
Clarifying nullness on keys

cdonnellytx commented:
Closing draft, going with https://github.com/dallasbeek/Dapper.Database/pull/23
